### PR TITLE
net-wireless/wireless-regdb: update 20230503, 20230213, 20220812

### PR DESCRIPTION
**Pull request by negril**
Implemented changes proposed in https://github.com/gentoo/gentoo/pull/31509#pullrequestreview-1484833642

Add blocker on net-wireless/crda, as it's no longer needed for >=kernel-4.15, and remove special treatment for it. See https://bugs.gentoo.org/693120.

Recompile regulatory.bin from db.txt.

Copy kernel config_check from net-wireless/iwd to ensure loading.

Bug: https://bugs.gentoo.org/898162

### CODE REVIEW

**Review by ZeroChaos- — CHANGES_REQUESTED** [review]
Check out what I did in the Pentoo overlay to maintain crda support please.

**Comment by negril** [comment]
Is there a use case for crda? If there is, I can add a use-flag for crda.

**Comment by thesamesam** [comment]
> Check out what I did in the Pentoo overlay to maintain crda support please.

You maintain crda in gentoo, is there a reason you didn't do it in ::gentoo?

**Comment by ZeroChaos-** [comment]
> > Check out what I did in the Pentoo overlay to maintain crda support please.
> 
> You maintain crda in gentoo, is there a reason you didn't do it in ::gentoo?

Yes, back when there were other maintainers they didn't like the idea.  That time has long passed so I see no reason not to move that work into Gentoo.  I would do it right now but I am away from my computer the next few days.  In the mean time I simply don't want this merged and to break the crda use cases.

**Comment by ZeroChaos-** [comment]
> Is there a use case for crda? If there is, I can add a use-flag for crda.

Yes, the use case for crda is that it can dynamically reload new wireless-regdb without rebuilding the whole kernel.  I know some folks change their kernel much more often than their regdb but there are use cases for rebuilding and updating the regdb which I would like to keep.  A hard blocker is a no, I am not even sure I am okay with a default off use flag unless the check for building the wireless regdb and key in as firmware actually works right.

**Comment by negril** [comment]
Sounds good enough for me. Will adjust.

**Comment by negril** [comment]
`net-wireless/iw` supports reloading the regulatory database as per https://patchwork.kernel.org/project/linux-wireless/patch/20180509153620.23630-1-seth.forshee@canonical.com/ with a call to `iw reg reload`.

So I don't see a need to support `net-wireless/crda`.

**Comment by ZeroChaos-** [comment]
I agree to a non-default use flag in that case. Thanks for finding that.  I still don't want a blocker though, crda is on most Gentoo systems and I will probably have to adjust other wireless things with this change.  I also respectfully request a bit of testing time to make sure the checks are all good and we don't need any other ones.  I am not home till Monday so some time this week I will have this tested.

**Comment by negril** [comment]
Further reading:

https://devel.fedoraproject.narkive.com/sIhqEvpX/fedora-crda-to-wireless-regdb-upgrade
https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1958918
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1003903

In short, crda was a helper util to load the regdb. It got replaced in 4.15 by in kernel facilites and is no longer useful because of it's limitations.

In short it's deprecated software should no longer be used, that needs to go and will be removed with the last kernels 4.14 LTS kernels in January 2024.

**Comment by ZeroChaos-** [comment]
@negril I'm back at my desk.  I really appreciate your work here.  I'm not going to let this drop, give me a little time to review this, because there are quite a few other packages.  All of this looks good, except the block on crda.

The blocker on crda makes no sense for a lot of reasons.  The main one being that the ebuilds are still installing files that only crda reads...  Please make a use flag for crda which removed the dep and doesn't install the files for crda when it's not set.  I am fine optionally removing the crda dep here, but making it a blocker will cause upgrade hiccups for no reason.

File: net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild [hunk 1]
```diff
@@ -3,8 +3,10 @@
 
 EAPI=8
 
+inherit linux-info
+
 MY_P="wireless-regdb-${PV:0:4}.${PV:4:2}.${PV:6:2}"
-DESCRIPTION="Binary regulatory database for CRDA"
+DESCRIPTION="Regulatory database for Linux"
```
Comment (ZeroChaos-) [inline net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild:9]: your version doesn't compile it either, so why are you changing this?

File: net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild [hunk 2]
```diff
@@ -13,25 +15,47 @@ LICENSE="ISC"
 SLOT="0"
 KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~ia64 ~loong ~mips ~ppc ~ppc64 ~riscv ~sparc ~x86"
 
+REQUIRED_USE="kernel_linux"
+
+pkg_pretend() {
+	if kernel_is -ge 4 15 && has_version net-wireless/crda; then
```
Comment (ZeroChaos-) [inline net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild:21]: I don't really have a problem with this, but it should probably go after the config checks and mention that the config needs to be correct to support not using crda.

File: net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild [hunk 3]
```diff
@@ -13,25 +15,47 @@ LICENSE="ISC"
...
+	fi
+}
+
 src_compile() {
-	einfo "Recompiling regulatory.bin from db.txt would break CRDA verify. Installing unmodified binary version."
+	true
 }
 
 src_install() {
+	if kernel_is -lt 4 15; then
```
Comment (ZeroChaos-) [inline net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild:47]: this is flat wrong.  The kernel checks above are not fatal (nor should they be) so users may be on a newer kernel which isn't configured properly and need crda.  I've said multiple times to add a crda use flag, I know you hate crda, but a non-default use flag is going to provide the best user experience here.

File: net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild [hunk 4]
```diff
@@ -13,25 +15,47 @@ LICENSE="ISC"
...
-	# can't find these files #643520
+	else
 	insinto /lib/firmware
-	doins regulatory.db
-	doins regulatory.db.p7s
-
-	doman regulatory.bin.5
+	doins regulatory.db regulatory.db.p7s
+	fi
+	doman -i18n= regulatory.bin.5 regulatory.db.5
```
Comment (ZeroChaos-) [inline net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild:59]: if we don't install regulatory.bin why are we installing regulatory.bin.5?  again, a crda use flag really makes sense here.

**Comment by ZeroChaos-** [comment]
per @thesamesam on irc there was some confusion with an earlier statement I made so I will clarify here.

Please introduce a use flag for crda.  It should be disabled by default.  When it is disabled, crda is not pulled in as a dep, and files needed by crda should not be installed.  DO NOT add a blocker against crda, it's an easier upgrade path if we simply allow it to get depcleaned.  Other packages still depend on crda, and while I'm willing to change that it would complicate the upgrade path to add the blocker at this time.

**Comment by ZeroChaos-** [comment]
I just took a look at net-wireless/iwd and some of the warning in there might make sense.  If we can use this work as a "let's get rid of crda" plan I'd like the checks to be as complete as possible so we can copy and paste them into the other ebuilds to improve checking and make sure users are good.

File: net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild [hunk 5]
```diff
@@ -3,35 +3,65 @@
...
 KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~ia64 ~loong ~mips ~ppc ~ppc64 ~riscv ~sparc ~x86"
+IUSE="crda"
+
+RDEPEND="crda? ( net-wireless/crda )"
+
+REQUIRED_USE="kernel_linux"
+
+pkg_pretend() {
+	if kernel_is -ge 4 15; then
+	if linux_config_exists && linux_chkconfig_builtin CFG80211 &&
```
Comment (ZeroChaos-) [inline net-wireless/wireless-regdb/wireless-regdb-20230503.ebuild:25]: we should also add a warning for when CFG80211 isn't selected at all.

**Pull request closed without merging**